drivers/espi/espi_emul.c

```diff
@@ -0,0 +1,208 @@
+
+	return (data->cfg.channel_caps & ch);
+}
+
+static int espi_emul_send_vwire(const struct device *dev, enum espi_vwire_signal vw, uint8_t level)
+{
+	const struct emul_espi_device_api *api;
+	struct espi_emul *emul;
+	struct espi_emul_data *data = dev->data;
+
+	if (!(data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
+		return -EIO;
+	}
+
+	emul = espi_emul_find(dev, EMUL_ESPI_HOST_CHIPSEL);
+
+	if (!emul) {
+		LOG_DBG("espi_emul not found");
+		return -EIO;
+	}
+
+	__ASSERT_NO_MSG(emul->api);
+	__ASSERT_NO_MSG(emul->api->set_vw);
+	api = emul->api;
+
+	return api->set_vw(emul, vw, level);
+}
+
+static int espi_emul_receive_vwire(const struct device *dev, enum espi_vwire_signal vw, uint8_t *level)
+{
+	const struct emul_espi_device_api *api;
+	struct espi_emul *emul;
+	struct espi_emul_data *data = dev->data;
+
+	if (!(data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
+		return -EIO;
+	}
+
+	emul = espi_emul_find(dev, EMUL_ESPI_HOST_CHIPSEL);
+
+	if (!emul) {
+		LOG_INF("espi_emul not found");
+		return -EIO;
+	}
+
+	__ASSERT_NO_MSG(emul->api);
+	__ASSERT_NO_MSG(emul->api->get_vw);
+	api = emul->api;
+
+	return api->get_vw(emul, vw, level);
+}
+
+static int espi_emul_manage_callback(const struct device *dev, struct espi_callback *callback, bool set)
+{
+	struct espi_emul_data *data = dev->data;
+
+	return espi_manage_callback(&data->callbacks, callback, set);
+}
+
+/**
+ * Set up a new emulator and add it to the list
+ *
+ * @param dev eSPI emulation controller device
+ */
+static int espi_emul_init(const struct device *dev)
+{
+	struct espi_emul_data *data = dev->data;
+	const struct emul_list_for_bus *list = dev->config;
+
+	sys_slist_init(&data->emuls);
+
+	return emul_init_for_bus_from_list(dev, list);
+}
+
+int espi_emul_register(const struct device *dev, const char *name,
+		       struct espi_emul *emul)
+{
+	struct espi_emul_data *data = dev->data;
+
+	sys_slist_append(&data->emuls, &emul->node);
+
+	LOG_INF("Register emulator '%s' at cs %u\n", name, emul->chipsel);
+
+	return 0;
+}
+
+/* Device instantiation */
+static struct emul_espi_driver_api emul_espi_driver_api = {
+	.espi_api = {
+		.config = espi_emul_config,
+		.get_channel_status = espi_emul_get_channel_status,
+		.send_vwire = espi_emul_send_vwire,
+		.receive_vwire = espi_emul_receive_vwire,
+		.manage_callback = espi_emul_manage_callback
+	},
+	.triger_event = emul_espi_triger_event,
+	.find_emul = espi_emul_find,
+};
+
+
+#define EMUL_LINK_AND_COMMA(node_id) {	    \
+		.label = DT_LABEL(node_id), \
+},
+
+#define ESPI_EMUL_INIT(n)					      \
+	static const struct emul_link_for_bus emuls_##n[] = {	      \
+		DT_FOREACH_CHILD(DT_DRV_INST(0), EMUL_LINK_AND_COMMA) \
+	};							      \
+	static struct emul_list_for_bus espi_emul_cfg_##n = {	      \
+		.children = emuls_##n,				      \
+		.num_children = ARRAY_SIZE(emuls_##n),		      \
+	};							      \
+	static struct espi_emul_data espi_emul_data_##n;	      \
+	DEVICE_AND_API_INIT(espi_##n,				      \
```

Comment:
Please switch to `DEVICE_DT_INST_DEFINE`.  See #31128.

Reply:
Done.